### PR TITLE
Create header files for all C++ classes

### PR DESCRIPTION
**Pull request by Nayef211**
## Description
- Creating header files for all C++ classes
- Moving all torchtext binding registrations to a separate file

### CODE REVIEW

File: torchtext/csrc/vectors.cpp [hunk 1]
```diff
@@ -203,7 +181,7 @@ void parse_chunk(const std::string &file_path, size_t offset,
 
 std::tuple<IndexMap, StringList>
 _concat_vectors(std::vector<std::shared_ptr<StringList>> chunk_tokens,
-                int64_t num_header_lines, int64_t num_lines) {
+                const int64_t num_header_lines, const int64_t num_lines) {
```
Comment (zhangguanheng66) [inline torchtext/csrc/vectors.cpp:184]: we are not passing by reference. Do we still need `const` here?

File: torchtext/csrc/vectors.cpp [hunk 2]
```diff
@@ -11,101 +11,79 @@
 #include <stdexcept>
```
Comment (zhangguanheng66) [inline torchtext/csrc/vectors.cpp:11]: Why not move those header files to `vectors.h` together

File: torchtext/csrc/vectors.cpp (hunk 1 above)
Comment (Nayef211) [reply torchtext/csrc/vectors.cpp:184]: Good call, I can remove this

File: torchtext/csrc/vocab.cpp [hunk 3]
```diff
@@ -1,119 +1,106 @@
+#include <ATen/Parallel.h>
 #include <stdexcept>
```
Comment (zhangguanheng66) [inline torchtext/csrc/vocab.cpp:2]: Same question about the header file.

File: torchtext/csrc/vectors.cpp (hunk 2 above)
Comment (Nayef211) [reply torchtext/csrc/vectors.cpp:11]: I think in general, it is good practice to not include gratuitious `#include` directives. A header file should not be included in a `.h` file if the `.h` file itself doesn't need those other header files.

@cpuhrsch can correct me if I am wrong

File: torchtext/csrc/vocab.cpp (hunk 3 above)
Comment (Nayef211) [reply torchtext/csrc/vocab.cpp:2]: Same as above ^

File: torchtext/csrc/vectors.cpp (hunk 2 above)
Comment (zhangguanheng66) [reply torchtext/csrc/vectors.cpp:11]: OK, thanks for the explanation.

File: torchtext/csrc/vectors.cpp (hunk 1 above)
Comment (mthrok) [reply torchtext/csrc/vectors.cpp:184]: If it's `const`, it's better to be annotated as `const`. Compilers might find an opportunity to optimize something. Also that will prevent accidental overwrite.

File: torchtext/csrc/register_bindings.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
+#include <regex_tokenizer.h>
+#include <sentencepiece.h>
+#include <torch/script.h>
+#include <vectors.h>
+#include <vocab.h>
+
+namespace torchtext {
+
```
Comment (mthrok) [inline torchtext/csrc/register_bindings.cpp:8]: I think now that the official documentation is updated to use `TORCH_LIBRARY` macro, which makes the registration cleaner and simpler, we can also migrate to this new approach, unless there is some issue about it. Did you try it?

File: torchtext/csrc/vectors.cpp (hunk 1 above)
Comment (Nayef211) [reply torchtext/csrc/vectors.cpp:184]: Readded

File: torchtext/csrc/vectors.cpp (hunk 2 above)
Comment (mthrok) [reply torchtext/csrc/vectors.cpp:11]: > I think in general, it is good practice to not include gratuitious `#include` directives. A header file should not be included in a `.h` file if the `.h` file itself doesn't need those other header files.
> 
> @cpuhrsch can correct me if I am wrong

I second that.

File: torchtext/csrc/register_bindings.cpp (hunk 4 above)
Comment (Nayef211) [reply torchtext/csrc/register_bindings.cpp:8]: Will do this in a follow up PR!

**Pull request merged**